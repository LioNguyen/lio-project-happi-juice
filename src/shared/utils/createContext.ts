import type React from "react";
import { createContext as originCreateContext, useContext } from "react";

function createContext<T>(): [React.Provider<T>, () => T, React.Context<T>] {
  const context = originCreateContext<T>({} as T);

  return [context.Provider, () => useContext(context), context];
}

export { createContext };
