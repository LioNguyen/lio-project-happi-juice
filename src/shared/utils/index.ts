export * from './lazyLoad'
export * from './tailwind'
