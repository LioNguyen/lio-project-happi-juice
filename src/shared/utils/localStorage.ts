/**
 * Save data to localStorage
 * @param key - The key under which the data will be stored
 * @param value - The data to store, can be any serializable value
 */
function setLocalStorage<T>(key: string, value: T): void {
  try {
    const serializedValue =
      typeof value === 'string' ? value : JSON.stringify(value)

    localStorage.setItem(key, serializedValue)
  } catch (error) {
    console.error(`Error saving ${key} to localStorage:`, error)
  }
}

/**
 * Retrieve data from localStorage
 * @param key - The key of the stored data
 * @returns The parsed data or null if not found or error occurs
 */
function getLocalStorage<T>(key: string): T | null {
  try {
    const serializedValue = localStorage.getItem(key)

    if (!serializedValue || serializedValue === 'undefined') {
      return null
    }

    // Try to parse JSON
    try {
      return JSON.parse(serializedValue) as T
    } catch {
      // Return as string if JSON parsing fails
      return serializedValue as T
    }
  } catch (error) {
    console.error(`Error reading ${key} from localStorage:`, error)
    return null
  }
}

/**
 * Remove data from localStorage
 * @param key - The key to remove from localStorage
 */
function removeLocalStorage(key: string): void {
  try {
    localStorage.removeItem(key)
  } catch (error) {
    console.error(`Error removing ${key} from localStorage:`, error)
  }
}

/**
 * Clear all data from localStorage
 */
function clearLocalStorage(): void {
  try {
    localStorage.clear()
  } catch (error) {
    console.error('Error clearing localStorage:', error)
  }
}

export {
  clearLocalStorage,
  getLocalStorage,
  removeLocalStorage,
  setLocalStorage,
}
