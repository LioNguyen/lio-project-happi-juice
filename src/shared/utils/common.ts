import { format, isValid, Locale, parseISO } from 'date-fns'
import { enUS, vi } from 'date-fns/locale'
import _ from 'lodash'

import { IOrderItem } from '@/domains/order'
import { ISelectOption } from '@/shared/types'

interface INumberFormatOptions {
  decimals?: number
  thousandSeparator?: string
  decimalSeparator?: string
}

// ==================== Date Utilities ====================
/**
 * Parse various date formats into Date object
 * @category Date
 * @param {Date | string | undefined} value - Input date
 * @returns {Date | undefined} Parsed Date object or undefined if invalid
 * @example
 * parseDate("2024-02-26") // Date object
 * parseDate("invalid") // undefined
 */
function parseDate(value: Date | string | undefined): Date | undefined {
  if (!value) return undefined
  if (value instanceof Date) return isValid(value) ? value : undefined

  try {
    // Try ISO string parsing first
    const date = parseISO(value)
    if (isValid(date)) return date

    // Fallback to direct Date creation
    const normalDate = new Date(value)
    return isValid(normalDate) ? normalDate : undefined
  } catch {
    return undefined
  }
}

/**
 * Format a date string according to specified format
 * @category Date
 * @param {string} date - Date string to format
 * @param {string} [dateFormat] - Output format (defaults to "yyyy-MM-dd")
 * @returns {string} Formatted date string
 * @example
 * formatDateTime("2024-02-26") // "2024-02-26"
 * formatDateTime("2024-02-26", "dd/MM/yyyy") // "26/02/2024"
 */
function formatDateTime(date: string, dateFormat?: string) {
  return format(date, dateFormat || 'yyyy-MM-dd')
}

const localeMap: Record<string, Locale> = {
  vi: vi,
  en: enUS,
}

// ==================== Number & Price Utilities ====================
/**
 * Format a number with customizable separators and decimals
 * @category Number
 * @param {number} value - Number to format
 * @param {INumberFormatOptions} [options] - Formatting options
 * @returns {string} Formatted number string
 */
const formatNumber = (
  value: number,
  options?: INumberFormatOptions,
): string => {
  const {
    decimals = 0,
    thousandSeparator = ',',
    decimalSeparator = '.',
  } = options || {}

  const parts = value.toFixed(decimals).split('.')
  parts[0] = parts[0].replace(/\B(?=(\d{3})+(?!\d))/g, thousandSeparator)

  return parts.join(decimalSeparator)
}

/**
 * Formats the price with thousand separators using locale-specific formatting
 * @category Price
 * @param {string} price - Price string to format
 * @returns {string} Formatted price string
 */
const formatPrice = (price: string) => {
  return Number(price).toLocaleString()
}

// ==================== Array Utilities ====================
/**
 * Transforms an array of select options into a sorted, joined string
 * @category Array
 * @param {ISelectOption[]} array - Array of select options
 * @param {string} [key] - Key to extract from each option
 * @param {string} [separator] - Separator for joining values
 * @returns {string} Sorted and joined string
 */
const transformToSortedString = (
  array: ISelectOption[],
  key: string = 'value',
  separator: string = ',',
): string => {
  return _.chain(array).map(key).sortBy().join(separator).value()
}

// ==================== Order Utilities ====================
/**
 * Calculates the total price for an array of order items
 * @category Order
 * @param {IOrderItem[]} items - Array of order items
 * @returns {number} Total price
 */
const calculateTotalPrice = (items: IOrderItem[]) => {
  return items.reduce((sum, item) => sum + item.price * item.quantity, 0)
}

export {
  // Date utils
  parseDate,
  formatDateTime,
  localeMap,

  // Number & Price utils
  formatNumber,
  formatPrice,

  // Array utils
  transformToSortedString,
  // Order utils
  calculateTotalPrice,
}
