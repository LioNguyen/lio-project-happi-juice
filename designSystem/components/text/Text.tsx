import { Slot } from "@radix-ui/react-slot";
import { cva, type VariantProps } from "class-variance-authority";
import * as React from "react";

import { cn } from "@/shared/utils";

const textVariants = cva("", {
  variants: {
    variant: {
      // Heading variants
      "heading-1": "text-text-primary text-xl lg:text-2xl font-semibold tracking-tight",

      // Body text variants
      default: "text-text-primary",
      secondary: "text-text-muted-foreground",
      description: "text-text-muted-foreground text-sm",

      // Form related variants
      label: "text-sm font-medium text-text-primary",
      error: "text-sm text-error",
      required: "text-error",

      // Interactive text
      link: "text-text-muted-foreground hover:text-primary hover:underline cursor-pointer",
    },
  },
  defaultVariants: {
    variant: "default",
  },
});

interface ITextProps extends React.HTMLAttributes<HTMLSpanElement>, VariantProps<typeof textVariants> {
  as?: React.ElementType;
  asChild?: boolean;
  htmlFor?: string;
}

const Text = React.forwardRef<HTMLElement, ITextProps>(
  ({ className, variant, as, asChild = false, children, ...props }, ref) => {
    const Comp = asChild ? Slot : as || "span";

    return (
      <Comp ref={ref} className={cn(textVariants({ variant, className }))} {...props}>
        {children}
      </Comp>
    );
  },
);

Text.displayName = "Text";

export default Text;
export { textVariants };
export type { ITextProps };
