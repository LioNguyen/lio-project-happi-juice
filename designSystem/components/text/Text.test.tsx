import { screen } from "@testing-library/react";
import { VariantProps } from "class-variance-authority";
import { ComponentPropsWithRef, createRef } from "react";
import { describe, expect, it } from "vitest";

import { renderWithTheme } from "@/shared/utils";
import Text, { textVariants } from "./Text";

type TextPropsWithRef = ComponentPropsWithRef<typeof Text>;
type TextVariantProps = VariantProps<typeof textVariants>;
type TextVariant = NonNullable<TextVariantProps["variant"]>;

const defaultProps: Partial<TextPropsWithRef> = {
  variant: "default",
};

const renderComponent = (props: Partial<TextPropsWithRef> = {}) => {
  return renderWithTheme(<Text {...defaultProps} {...props} />);
};

describe("Text Component", () => {
  // Snapshot Tests
  it("should match snapshot with default props", () => {
    const { container } = renderComponent({ children: "Text content" });
    expect(container.firstChild).toMatchSnapshot();
  });

  // Test render
  it("should render text with default props", () => {
    renderComponent({ children: "Hello world" });

    const text = screen.getByText("Hello world");
    expect(text).toBeInTheDocument();
    expect(text.tagName.toLowerCase()).toBe("span");
    expect(text).toHaveClass("text-text-primary");
  });

  // Test className custom
  it("should merge custom className with default classes", () => {
    renderComponent({ className: "custom-class", children: "Text" });

    const text = screen.getByText("Text");
    expect(text).toHaveClass("custom-class");
    expect(text).toHaveClass("text-text-primary");
  });

  // Test asChild prop
  it("should render as child component when asChild is true", () => {
    renderComponent({ asChild: true, children: <a href="#">Link Text</a> });

    const link = screen.getByRole("link");
    expect(link).toBeInTheDocument();
    expect(link).toHaveClass("text-text-primary");
  });

  // Test 'as' prop
  it("should render as different HTML element when 'as' prop is provided", () => {
    renderComponent({ as: "p", children: "Paragraph text" });

    const text = screen.getByText("Paragraph text");
    expect(text.tagName.toLowerCase()).toBe("p");
  });

  // Test htmlFor attribute for label variant
  it("should apply htmlFor attribute when variant is label", () => {
    renderComponent({
      variant: "label",
      htmlFor: "input-id",
      children: "Label text",
    });

    const label = screen.getByText("Label text");
    expect(label).toHaveAttribute("for", "input-id");
  });

  // Test ref forwarding
  it("should forward ref correctly", () => {
    const ref = createRef<HTMLSpanElement>();
    renderComponent({ ref, children: "Text with ref" });

    const text = screen.getByText("Text with ref");
    expect(ref.current).toBe(text);
  });

  // Test variants
  describe("Variants", () => {
    const variants: TextVariant[] = [
      "heading-1",
      "default",
      "secondary",
      "description",
      "label",
      "error",
      "required",
      "link",
    ] as const;

    variants.forEach(variant => {
      it(`should render ${variant} variant correctly`, () => {
        renderComponent({ variant, children: `${variant} text` });

        const text = screen.getByText(`${variant} text`);
        switch (variant) {
          case "heading-1":
            expect(text).toHaveClass("text-text-primary text-xl lg:text-2xl font-semibold tracking-tight");
            break;
          case "default":
            expect(text).toHaveClass("text-text-primary");
            break;
          case "secondary":
            expect(text).toHaveClass("text-text-muted-foreground");
            break;
          case "description":
            expect(text).toHaveClass("text-text-muted-foreground text-sm");
            break;
          case "label":
            expect(text).toHaveClass("text-sm font-medium text-text-primary");
            break;
          case "error":
            expect(text).toHaveClass("text-sm text-error");
            break;
          case "required":
            expect(text).toHaveClass("text-error");
            break;
          case "link":
            expect(text).toHaveClass("text-text-muted-foreground hover:text-primary hover:underline cursor-pointer");
            break;
        }
      });
    });
  });
});
