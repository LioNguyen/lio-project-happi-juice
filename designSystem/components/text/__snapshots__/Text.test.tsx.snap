// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Text Component > should match snapshot with default props 1`] = `
<span
  class="text-text-primary"
>
  Text content
</span>
`;
